services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DefaultS3TransferManager.java

```diff
@@ -29,79 +34,172 @@
+            PutObjectRequest putObjectRequest = uploadRequest.putObjectRequest();
+            AsyncRequestBody requestBody = requestBodyFor(uploadRequest);
 
-        CompletableFuture<CompletedUpload> future = putObjFuture.thenApply(r -> DefaultCompletedUpload.builder()
-                                                                                                      .response(r)
-                                                                                                      .build());
-        return new DefaultUpload(CompletableFutureUtils.forwardExceptionTo(future, putObjFuture));
+            CompletableFuture<PutObjectResponse> putObjFuture = s3CrtAsyncClient.putObject(putObjectRequest, requestBody);
+
+            CompletableFuture<CompletedUpload> future = putObjFuture.thenApply(r -> CompletedUpload.builder()
+                                                                                                   .response(r)
+                                                                                                   .build());
+            return new DefaultUpload(CompletableFutureUtils.forwardExceptionTo(future, putObjFuture));
+        } catch (Throwable throwable) {
+            return new DefaultUpload(CompletableFutureUtils.failedFuture(throwable));
+        }
     }
 
     @Override
-    public Download download(DownloadRequest downloadRequest) {
-        CompletableFuture<GetObjectResponse> getObjectFuture =
-            s3CrtAsyncClient.getObject(downloadRequest.getObjectRequest(),
-                                       AsyncResponseTransformer.toFile(downloadRequest.destination()));
-        CompletableFuture<CompletedDownload> future =
-            getObjectFuture.thenApply(r -> DefaultCompletedDownload.builder().response(r).build());
+    public UploadDirectoryTransfer uploadDirectory(UploadDirectoryRequest uploadDirectoryRequest) {
+        try {
+            Validate.paramNotNull(uploadDirectoryRequest, "uploadDirectoryRequest");
+            Path directory = uploadDirectoryRequest.sourceDirectory();
+
+            Validate.isTrue(Files.exists(directory), "The source directory (%s) provided does not exist", directory);
+            boolean followSymbolicLinks = transferConfiguration.resolveUploadDirectoryFollowSymbolicLinks(uploadDirectoryRequest);
+            if (followSymbolicLinks) {
+                Validate.isFalse(Files.isRegularFile(directory), "The source directory (%s) provided is not a "
+                                                                 + "directory", directory);
+            } else {
+                Validate.isFalse(Files.isRegularFile(directory, LinkOption.NOFOLLOW_LINKS), "The source directory (%s) provided"
+                                                                                            + " is not a "
+                                                                                            + "directory", directory);
+            }
```

Comment:
Because it's easy to do, should we do these file system checks in uploadDIrectoryManager.uploadDirectory, so that they can be done on the executor instead of the calling thread? It's probably negligible performance-wise for most applications, but something like an NFS could block the calling thread for a more noticeable amount of time.